iot-hub/Samples/device/DeviceReconnectionSample/readme.md

```diff
@@ -0,0 +1,56 @@
+# Device Reconnection Sample
+
+This sample code demonstrates the various connection status changes and connection status change reasons the Device Client can return, and how to handle them.
+
+The device client exhibits the following connection status changes with reason:
+
+<table>
+  <tr>
+    <th> Connection Status </th>
+    <th> Change Reason </th>
+    <th> Ownership of connectivity </th>
+    <th> Comments </th>
+  </tr>
+  <tr>
+    <td> Connected </td>
+    <td> Connection_Ok </td>
+    <td> SDK </td>
+    <td> SDK tries to remain connected to the service and can carry out all operations as normal. </td>
+  </tr>
+  <tr>
+    <td> Disconnected_Retrying </td>
+    <td> Communication_Error </td>
+    <td> SDK </td>
+    <td> When disconnection happens because of any reason (network failures, transient loss of connectivity etc.), SDK makes best attempt to connect back to IotHub. The RetryPolicy applied on the DeviceClient will be used to determine the count of reconnection attempts for <em>retriable</em> errors. </td>
+  </tr>
+  <tr>
+    <td rowspan="4"> Disconnected </td>
+    <td> Device_Disabled </td>
+    <td rowspan="4"> Application </td>
+    <td> This signifies that the device/ module has been deleted or marked as disabled (on your hub instance). <br/> Fix the device/ module status in Azure before attempting to create the associated client instance. </td>
+  </tr>
+  <tr>
+    <td> Bad_Credential </td>
+    <td> Supplied credential isn’t good for device to connect to service. <br/> Fix the supplied credentials before attempting to reconnect again </td>
+  </tr>
+  <tr>
+    <td> Communication_Error </td>
+    <td> This is the state when SDK landed up in a non-retriable error during communication. <br/> If you want to perform more operations on the device client, you should dispose and then re-initialize the client. </td>
```

Comment:
What should be done with pending commands, telemetry that didn't get confirmed as sent, etc.?

Reply:
The call wouldn't complete, so technically the only pending item is the operation that was being executed when the error happened. I'll add a note explaining that, sure.

Reply:
Presumably there could be lots of other pending actions right?

I run into a communication error at time T0, and my retries are exhausted at time T10. From T0 to T9 I had lots of initiated telemetry events and commands received. They are all pending due to the comm error that started at T0.

If I dispose my client and init, I've lost all that stuff right?

Reply:
Are we assuming that these telemetry operations were initiated by separate tasks, or by the same task? 
I'll sync up with you offline on this, I might be misinterpreting your question.